### PR TITLE
Linux: Build CMake if the checked out version is newer than the installed version.

### PR DESCRIPTION
**Pull request by spevans**
For Linux only, if the checked out CMake repository is a newer version than the installed CMake version, build and use CMake from source.

This does not affect macOS build  or set any minimum required CMake version in CMakeLists.txt

### CODE REVIEW

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test

File: utils/swift_build_support/swift_build_support/cmake.py [hunk 1]
```diff
@@ -177,3 +181,98 @@ def build_args(self):
...
         return build_args
+
+    # Determine the version of the installed CMake binary.
+    def installed_cmake_version(self, cmake_binary):
+        version = shell.capture([cmake_binary, '--version'], dry_run=False,
+                                    echo=True, optional=True)
+        if version is None:
+            return (0, 0, 0)
+
+        print(version)
```
Comment (compnerd) [inline utils/swift_build_support/swift_build_support/cmake.py]: debugging left overs?

File: utils/swift_build_support/swift_build_support/cmake.py [hunk 2]
```diff
@@ -177,3 +181,98 @@ def build_args(self):
...
+        x = re.findall(r'cmake version (\d+)\.(\d+)\.(\d+)', version.rstrip())
+        if len(x) == 1:
+            (c_major, c_minor, c_patch) = map(int, x[0])
+
+        return (c_major, c_minor, c_patch)
```
Comment (compnerd) [inline utils/swift_build_support/swift_build_support/cmake.py]: In the case that the findall fails, this is going to fail due to use of undefined variables.

File: utils/swift_build_support/swift_build_support/cmake.py [hunk 3]
```diff
@@ -177,3 +181,98 @@ def build_args(self):
...
+
+    # Determine the version of the checked out CMake source.
+    def cmake_source_version(self, cmake_source_dir):
+        cmake_version_file = os.path.join(cmake_source_dir, 'Source',
+                                              'CMakeVersion.cmake')
+        print("cmake_version_file: %s" % cmake_version_file)
```
Comment (compnerd) [inline utils/swift_build_support/swift_build_support/cmake.py]: debugging left over?

File: utils/swift_build_support/swift_build_support/cmake.py [hunk 4]
```diff
@@ -177,3 +181,98 @@ def build_args(self):
...
+        major = -1
+        minor = -1
+        patch = -1
+
+        file = open(cmake_version_file, "r")
+        for line in file.readlines():
+            m = re.findall(r'set\(CMake_VERSION_MAJOR (\d+)\)', line)
+            if len(m) == 1:
+                major = int(m[0])
+                next
```
Comment (compnerd) [inline utils/swift_build_support/swift_build_support/cmake.py]: I don't know this keyword.  What does it do?

File: utils/swift_build_support/swift_build_support/cmake.py [hunk 5]
```diff
@@ -177,3 +181,98 @@ def build_args(self):
...
+        cmake_bootstrap = os.path.join(source_root, 'cmake', 'bootstrap')
+
+        if hasattr(self.args, 'build_script_impl_args'):
+            for opt in self.args.build_script_impl_args:
+                m = re.findall('--build-dir=(.*)', opt)
+                if len(m) == 1:
+                    build_root = m[0]
+
+        cmake_build_dir = os.path.join(build_root, 'cmake-%s' % self.args.host_target)
+        shell.call_without_sleeping(['mkdir', '-p', cmake_build_dir], echo=True)
```
Comment (compnerd) [inline utils/swift_build_support/swift_build_support/cmake.py]: Please use `os.makedirs`.  This will fail on Windows.

**Comment by compnerd** [comment]
CC: @shahmishal @Rostepher @drodriguez

File: utils/swift_build_support/swift_build_support/cmake.py (hunk 5 above)
Comment (spevans) [reply utils/swift_build_support/swift_build_support/cmake.py]: `os.makedirs()` wasn't available using the default python installed on Ubuntu18.04. (Python 2.7.15rc1). Is the python version on Windows guaranteed to have it? If so I could use something like:
```
if platform.System() == 'Linux':
  mkdir -p
else
  os.makedirs
```

File: utils/swift_build_support/swift_build_support/cmake.py [hunk 6]
```diff
@@ -177,3 +181,95 @@ def build_args(self):
...
+                cmake_binary = self.args.cmake
+        except AttributeError:
+            cmake_binary = 'cmake'
+
+        (i_major, i_minor, i_patch) = self.installed_cmake_version(cmake_binary)
+        (s_major, s_minor, s_patch) = self.cmake_source_version(cmake_source_dir)
+        if (i_major > s_major or (i_major == s_major and i_minor >= s_minor) or
+            (i_major == s_major and i_minor == s_minor and i_patch >= s_patch)):
+            return
+        else:
```
Comment (compnerd) [inline utils/swift_build_support/swift_build_support/cmake.py]: else is unnecessary after the return.

**Review by compnerd — COMMENTED** [review]
LGTM

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by spevans** [comment]
@swift-ci test macOS

**Comment by spevans** [comment]
@swift-ci test macOS

**Comment by shahmishal** [comment]
@swift-ci test

**Comment by shahmishal** [comment]
@spevans can you please verify this works on Ubuntu 14.04 and 18.04?

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@shahmishal Ive just tested it locally on 14.04 and 18.04 and it worked ok.

File: utils/swift_build_support/swift_build_support/cmake.py [hunk 7]
```diff
@@ -177,3 +181,95 @@ def build_args(self):
...
+        try:
+            if self.args.cmake is not None:
+                cmake_binary = self.args.cmake
+        except AttributeError:
+            cmake_binary = 'cmake'
+
+        (i_major, i_minor, i_patch) = self.installed_cmake_version(cmake_binary)
+        (s_major, s_minor, s_patch) = self.cmake_source_version(cmake_source_dir)
+        if (i_major > s_major or (i_major == s_major and i_minor >= s_minor) or
+            (i_major == s_major and i_minor == s_minor and i_patch >= s_patch)):
```
Comment (drodriguez) [inline utils/swift_build_support/swift_build_support/cmake.py]: In Python you can simply do `self.installed_cmake_version(cmake_binary) > self.cmake_source_version(cmake_source_dir)` and the tuples will be compared lexicographically.

**Comment by shahmishal** [comment]
I am going to merge this, but please create a new PR to simplify the suggestion @drodriguez had. Thanks!

File: utils/swift_build_support/swift_build_support/cmake.py (hunk 7 above)
Comment (spevans) [reply utils/swift_build_support/swift_build_support/cmake.py]: Thanks for the info I will sort out a fix.

**Comment by aschwaighofer** [comment]
This breaks the incremental build because of python lint errors.
https://ci.swift.org/job/oss-swift-incremental-RA-osx/8587

I am going to revert.

**Comment by aschwaighofer** [comment]
```
FAIL: Swift(macosx-x86_64) :: Python/python-lint.swift (7244 of 12214)
******************** TEST 'Swift(macosx-x86_64) :: Python/python-lint.swift' FAILED ********************
Script:
--
: 'RUN: at line 1';   /usr/bin/python /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/utils/python_lint.py
--
Exit Code: 1

Command Output (stdout):
--
./utils/build-script:1046:80: E501 line too long (82 > 79 characters)
./utils/swift_build_support/swift_build_support/cmake.py:23:1: I100 Import statements are in the wrong order. import re should be before from .
./utils/swift_build_support/swift_build_support/cmake.py:24:1: I100 Import statements are in the wrong order. import os should be before import re
./utils/swift_build_support/swift_build_support/cmake.py:188:37: E127 continuation line over-indented for visual indent
./utils/swift_build_support/swift_build_support/cmake.py:191:80: E501 line too long (82 > 79 characters)
./utils/swift_build_support/swift_build_support/cmake.py:200:47: E127 continuation line over-indented for visual indent
./utils/swift_build_support/swift_build_support/cmake.py:224:36: E127 continuation line over-indented for visual indent
./utils/swift_build_support/swift_build_support/cmake.py:238:80: E501 line too long (86 > 79 characters)
./utils/swift_build_support/swift_build_support/cmake.py:246:41: E127 continuation line over-indented for visual indent
./utils/swift_build_support/swift_build_support/cmake.py:250:80: E501 line too long (82 > 79 characters)
./utils/swift_build_support/swift_build_support/cmake.py:251:80: E501 line too long (85 > 79 characters)
./utils/swift_build_support/swift_build_support/cmake.py:257:80: E501 line too long (80 > 79 characters)
./utils/swift_build_support/swift_build_support/cmake.py:268:80: E501 line too long (80 > 79 characters)
./utils/swift_build_support/swift_build_support/cmake.py:269:80: E501 line too long (81 > 79 characters)
./utils/swift_build_support/swift_build_support/cmake.py:271:13: E129 visually indented line with same indent as next logical line
./utils/swift_build_support/swift_build_support/cmake.py:271:80: E501 line too long (80 > 79 characters)

--

********************
```

**Comment by aschwaighofer** [comment]
Please reapply with those errors fixed.

**Comment by spevans** [comment]
@aschwaighofer  I'll sort out the linter errors, but shouldnt the linter have run as part of the macOS ci-tests and failed earlier or is it run as a separate CI test?

**Comment by aschwaighofer** [comment]
@spevans It should run as part of ``@swift-ci test`` (if not even smoke-test).

**Pull request merged**